ext/sodium/libsodium.c

```diff
@@ -0,0 +1,2682 @@
+							  &msg, &msg_len,
+							  &key, &key_len) == FAILURE) {
+		return;
+	}
+	if (key_len != crypto_shorthash_KEYBYTES) {
+		zend_throw_exception(sodium_exception_ce,
+				   "crypto_shorthash(): key size should be "
+				   "CRYPTO_SHORTHASH_KEYBYTES bytes",
+				   0);
+		return;
+	}
+	hash = zend_string_alloc(crypto_shorthash_BYTES, 0);
+	if (crypto_shorthash((unsigned char *) ZSTR_VAL(hash), msg,
+						 (unsigned long long) msg_len, key) != 0) {
+		zend_string_free(hash);
+		zend_throw_exception(sodium_exception_ce, "crypto_shorthash(): internal error", 0);
+		return;
+	}
+	ZSTR_VAL(hash)[crypto_shorthash_BYTES] = 0;
+
+	RETURN_STR(hash);
+}
+
+PHP_FUNCTION(sodium_crypto_secretbox)
+{
+	zend_string	  *ciphertext;
+	unsigned char *key;
+	unsigned char *msg;
+	unsigned char *nonce;
+	size_t		   key_len;
+	size_t		   msg_len;
+	size_t		   nonce_len;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "sss",
+							  &msg, &msg_len,
+							  &nonce, &nonce_len,
+							  &key, &key_len) == FAILURE) {
+		return;
+	}
+	if (nonce_len != crypto_secretbox_NONCEBYTES) {
+		zend_throw_exception(sodium_exception_ce,
+				   "crypto_secretbox(): nonce size should be "
+				   "CRYPTO_SECRETBOX_NONCEBYTES bytes",
+				   0);
+		return;
+	}
+	if (key_len != crypto_secretbox_KEYBYTES) {
+		zend_throw_exception(sodium_exception_ce,
+				   "crypto_secretbox(): key size should be "
+				   "CRYPTO_SECRETBOX_KEYBYTES bytes",
+				   0);
+		return;
+	}
+	if (SIZE_MAX - msg_len <= crypto_secretbox_MACBYTES) {
+		zend_throw_exception(sodium_exception_ce, "arithmetic overflow", 0);
+		return;
+	}
+	ciphertext = zend_string_alloc((size_t) msg_len + crypto_secretbox_MACBYTES, 0);
+	if (crypto_secretbox_easy((unsigned char *) ZSTR_VAL(ciphertext),
+							  msg, (unsigned long long) msg_len,
+							  nonce, key) != 0) {
+		zend_string_free(ciphertext);
+		zend_throw_exception(sodium_exception_ce, "crypto_secretbox(): internal error", 0);
+		return;
+	}
+	ZSTR_VAL(ciphertext)[msg_len + crypto_secretbox_MACBYTES] = 0;
+
+	RETURN_STR(ciphertext);
+}
+
+PHP_FUNCTION(sodium_crypto_secretbox_open)
+{
+	zend_string	  *msg;
+	unsigned char *key;
+	unsigned char *ciphertext;
+	unsigned char *nonce;
+	size_t		   key_len;
+	size_t		   ciphertext_len;
+	size_t		   nonce_len;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "sss",
+							  &ciphertext, &ciphertext_len,
+							  &nonce, &nonce_len,
+							  &key, &key_len) == FAILURE) {
+		return;
+	}
+	if (nonce_len != crypto_secretbox_NONCEBYTES) {
+		zend_throw_exception(sodium_exception_ce,
+				   "crypto_secretbox_open(): nonce size should be "
+				   "CRYPTO_SECRETBOX_NONCEBYTES bytes",
+				   0);
+		return;
+	}
+	if (key_len != crypto_secretbox_KEYBYTES) {
+		zend_throw_exception(sodium_exception_ce,
+				   "crypto_secretbox_open(): key size should be "
+				   "CRYPTO_SECRETBOX_KEYBYTES bytes",
+				   0);
+		return;
+	}
+	if (ciphertext_len < crypto_secretbox_MACBYTES) {
+		RETURN_FALSE;
```

Comment:
`strcmp` is a bad example, as it does not use exceptions in the first place, and opts for error codes. You are using exceptions on the other hand.

Exceptions are not meant to point out something that requires fixing, in that case an error is asked for (or even a fatal error, depending on the severity). Exceptions indicate that your routine was not able to perform its duty and uphold its promise, in this case decoding something. The caller can catch that exception and try to recover, according to her domain.

Your current API violates this and requires double error handling:

```php
try {
    if (sodium_crypto_secretbox_open('foo', 'bar', 'baz') === false) {
        // could not open, handle problem ...
    } else {
        // successfully opened, continue flow ...
    }
} catch (SodiumException $e) {
    // Actually an InvalidArgumentException ...
}
```

You might ask now why `InvalidArgumentException` is an exception and not an error. We cannot tell whether it was a user-level developer mistake or end-user supplied data that lead to the problem, hence, we throw an exception which allows the user-level developer to take appropriate action. This could be an error message that is displayed to the end-user.